File: app-ios-nativa/Dominio/Pruebas/PruebasDominio/PruebasBiblioteca.swift
import Foundation
import Testing

@testable import Dominio

private func elemento(
    _ id: String,
    url: String = "https://a.com",
    titulo: String? = nil,
    etiquetas: [String] = []
) -> Elemento {
    Elemento(id: id, url: url, titulo: titulo, etiquetas: etiquetas, actualizadoEn: 100)
}

@Suite("Buscar en la lista")
struct PruebasBuscar {
    private let a = elemento("a", url: "https://python.org", titulo: "Documentación Python")
    private let b = elemento(
        "b", url: "https://otra.com", titulo: "Otra cosa", etiquetas: ["python"])
    private let c = elemento("c", url: "https://nada.com", titulo: "Nada que ver")

    @Test("busca en título, URL y etiquetas, sin distinguir mayúsculas")
    func buscaEnLosTresSitios() {
        #expect(Biblioteca.buscar([a, b, c], texto: "PYTHON").map(\.id) == ["a", "b"])
    }

    @Test("sin texto no filtra")
    func sinTexto() {
        #expect(Biblioteca.buscar([a, b, c], texto: "").count == 3)
        #expect(Biblioteca.buscar([a, b, c], texto: "   ").count == 3)
    }

    @Test("lo que no está no aparece")
    func sinResultados() {
        #expect(Biblioteca.buscar([a, b, c], texto: "no-existe").isEmpty)
    }
}

@Suite("Etiquetas en uso")
struct PruebasEtiquetasEnUso {
    @Test("las distintas, ordenadas y sin repetir")
    func distintasOrdenadas() {
        let a = elemento("a", etiquetas: ["ocio", "pendiente"])
        let b = elemento("b", etiquetas: ["trabajo", "ocio"])
        let c = elemento("c")

        #expect(Biblioteca.etiquetasEnUso([a, b, c]) == ["ocio", "pendiente", "trabajo"])
    }

    @Test("sin elementos, o sin etiquetas, no hay ninguna")
    func vacio() {
        #expect(Biblioteca.etiquetasEnUso([]).isEmpty)
        #expect(Biblioteca.etiquetasEnUso([elemento("a")]).isEmpty)
    }

    @Test("cuenta cuántos enlaces lleva cada una")
    func recuento() {
        let a = elemento("a", etiquetas: ["ocio", "trabajo"])
        let b = elemento("b", etiquetas: ["ocio"])
        let c = elemento("c")

        #expect(Biblioteca.recuentoPorEtiqueta([a, b, c]) == ["ocio": 2, "trabajo": 1])
    }
}

@Suite("Filtrar por etiqueta")
struct PruebasFiltrarPorEtiqueta {
    private let a = elemento("a", etiquetas: ["ocio"])
    private let b = elemento("b", etiquetas: ["trabajo"])

    @Test("deja los que la llevan")
    func filtra() {
        #expect(Biblioteca.filtrarPorEtiqueta([a, b], etiqueta: "ocio").map(\.id) == ["a"])
    }

    @Test("sin etiqueta elegida no filtra")
    func sinEtiqueta() {
        #expect(Biblioteca.filtrarPorEtiqueta([a, b], etiqueta: nil).count == 2)
        #expect(Biblioteca.filtrarPorEtiqueta([a, b], etiqueta: "").count == 2)
    }
}

@Suite("Renombrar y quitar una etiqueta")
struct PruebasRenombrarEtiqueta {
    @Test("devuelve solo los enlaces que cambian, con la fecha movida")
    func soloLosQueCambian() {
        let a = elemento("a", etiquetas: ["ocio"])
        let b = elemento("b", etiquetas: ["trabajo"])

        let cambiados = Biblioteca.renombrarEtiqueta(
            en: [a, b],
            vieja: "ocio",
            nueva: "tiempo libre",
            ahora: relojFijo(200)
        )

        #expect(cambiados.map(\.id) == ["a"])
        #expect(cambiados[0].etiquetas == ["tiempo libre"])
        #expect(cambiados[0].actualizadoEn == 200)
    }

    @Test("renombrar a una que el enlace ya tenía las funde, no la repite")
    func fusionSinRepetir() {
        let a = elemento("a", etiquetas: ["ocio", "casa"])

        let cambiados = Biblioteca.renombrarEtiqueta(en: [a], vieja: "ocio", nueva: "casa")

        #expect(cambiados[0].etiquetas == ["casa"])
    }

    @Test("quitar una etiqueta deja el resto en paz")
    func quitarUna() {
        let a = elemento("a", etiquetas: ["ocio", "casa"])
        let b = elemento("b", etiquetas: ["trabajo"])

        let cambiados = Biblioteca.quitarEtiqueta(
            en: [a, b], etiqueta: "ocio", ahora: relojFijo(200))

        #expect(cambiados.map(\.id) == ["a"])
        #expect(cambiados[0].etiquetas == ["casa"])
        #expect(cambiados[0].actualizadoEn == 200)
    }
}

File: app-android/dominio/src/test/kotlin/com/jmortizsilva/guardarenlaces/dominio/PruebasBiblioteca.kt
package com.jmortizsilva.guardarenlaces.dominio

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private fun elemento(
    id: String,
    url: String = "https://a.com",
    titulo: String? = null,
    etiquetas: List<String> = emptyList(),
) = Elemento(id = id, url = url, titulo = titulo, etiquetas = etiquetas, actualizadoEn = 100)

class PruebasBuscar {
    private val a = elemento("a", url = "https://python.org", titulo = "Documentación Python")
    private val b =
        elemento("b", url = "https://otra.com", titulo = "Otra cosa", etiquetas = listOf("python"))
    private val c = elemento("c", url = "https://nada.com", titulo = "Nada que ver")

    @Test
    fun `busca en titulo, URL y etiquetas, sin distinguir mayusculas`() {
        assertEquals(listOf("a", "b"), Biblioteca.buscar(listOf(a, b, c), "PYTHON").map { it.id })
    }

    @Test
    fun `sin texto no filtra`() {
        assertEquals(3, Biblioteca.buscar(listOf(a, b, c), "").size)
        assertEquals(3, Biblioteca.buscar(listOf(a, b, c), "   ").size)
    }

    @Test
    fun `lo que no esta no aparece`() {
        assertTrue(Biblioteca.buscar(listOf(a, b, c), "no-existe").isEmpty())
    }
}

class PruebasEtiquetasEnUso {
    @Test
    fun `las distintas, ordenadas y sin repetir`() {
        val a = elemento("a", etiquetas = listOf("ocio", "pendiente"))
        val b = elemento("b", etiquetas = listOf("trabajo", "ocio"))
        val c = elemento("c")

        assertEquals(
            listOf("ocio", "pendiente", "trabajo"),
            Biblioteca.etiquetasEnUso(listOf(a, b, c)),
        )
    }

    @Test
    fun `se ordenan como en un diccionario, con las tildes en su sitio`() {
        // Comparando códigos de letra, «ábaco» iría detrás de «zumo» y «Casa» delante de «barco».
        val a = elemento("a", etiquetas = listOf("zumo", "ábaco", "Casa", "barco"))

        assertEquals(listOf("ábaco", "barco", "Casa", "zumo"), Biblioteca.etiquetasEnUso(listOf(a)))
    }

    @Test
    fun `las disponibles suman las reservadas, sin repetir y en orden`() {
        val a = elemento("a", etiquetas = listOf("ocio", "trabajo"))

        assertEquals(
            listOf("casa", "ocio", "trabajo"),
            Biblioteca.etiquetasDisponibles(listOf(a), reservadas = listOf("ocio", "casa")),
        )
    }

    @Test
    fun `sin elementos, o sin etiquetas, no hay ninguna`() {
        assertTrue(Biblioteca.etiquetasEnUso(emptyList()).isEmpty())
        assertTrue(Biblioteca.etiquetasEnUso(listOf(elemento("a"))).isEmpty())
    }

    @Test
    fun `cuenta cuantos enlaces lleva cada una`() {
        val a = elemento("a", etiquetas = listOf("ocio", "trabajo"))
        val b = elemento("b", etiquetas = listOf("ocio"))
        val c = elemento("c")

        assertEquals(
            mapOf("ocio" to 2, "trabajo" to 1),
            Biblioteca.recuentoPorEtiqueta(listOf(a, b, c)),
        )
    }
}

class PruebasFiltrarPorEtiqueta {
    private val a = elemento("a", etiquetas = listOf("ocio"))
    private val b = elemento("b", etiquetas = listOf("trabajo"))

    @Test
    fun `deja los que la llevan`() {
        assertEquals(listOf("a"), Biblioteca.filtrarPorEtiqueta(listOf(a, b), "ocio").map { it.id })
    }

    @Test
    fun `sin etiqueta elegida no filtra`() {
        assertEquals(2, Biblioteca.filtrarPorEtiqueta(listOf(a, b), null).size)
        assertEquals(2, Biblioteca.filtrarPorEtiqueta(listOf(a, b), "").size)
    }
}

class PruebasRenombrarEtiqueta {
    @Test
    fun `devuelve solo los enlaces que cambian, con la fecha movida`() {
        val a = elemento("a", etiquetas = listOf("ocio"))
        val b = elemento("b", etiquetas = listOf("trabajo"))

        val cambiados =
            Biblioteca.renombrarEtiqueta(listOf(a, b), "ocio", "tiempo libre", relojFijo(200))

        assertEquals(listOf("a"), cambiados.map { it.id })
        assertEquals(listOf("tiempo libre"), cambiados[0].etiquetas)
        assertEquals(200, cambiados[0].actualizadoEn)
    }

    @Test
    fun `renombrar a una que el enlace ya tenia las funde, no la repite`() {
        val a = elemento("a", etiquetas = listOf("ocio", "casa"))

        val cambiados = Biblioteca.renombrarEtiqueta(listOf(a), "ocio", "casa")

        assertEquals(listOf("casa"), cambiados[0].etiquetas)
    }

    @Test
    fun `quitar una etiqueta deja el resto en paz`() {
        val a = elemento("a", etiquetas = listOf("ocio", "casa"))
        val b = elemento("b", etiquetas = listOf("trabajo"))

        val cambiados = Biblioteca.quitarEtiqueta(listOf(a, b), "ocio", relojFijo(200))

        assertEquals(listOf("a"), cambiados.map { it.id })
        assertEquals(listOf("casa"), cambiados[0].etiquetas)
        assertEquals(200, cambiados[0].actualizadoEn)
    }
}
